### PR TITLE
[daikinairbase] Daikin Airbase Binding initial contribution

### PR DESCRIPTION
**Pull request by psmedley**
[daikinairbase] Initial Contribution - Daikin Airbase controller binding

New binding for Daikin Airbase controllers. Unfortunately Daikin Australia made the API different to the API used for the daikin binding. I also intend to add support for zones in the fullness of time that is not supported by the current daikin binding.

### CODE REVIEW

**Comment by psmedley** [comment]
Hi - what can I do to help get this merged?

File: bundles/org.openhab.binding.daikinairbase/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+The Daikin binding allows you to control your Daikin Airbase air conditioning units with openHAB. In order to do so, your Daikin air conditioning unit must have a BRP15B61 WiFi adapter (Also known as Airbase) installed.
+
+## Supported Things
+
+Daikin air conditioning units with a BRP15B61 (also known as an Airbase)installed. 
+
+## Discovery
+
+This addon will broadcast messages on your local network looking for Daikin Airbase air conditioning units and adding them to the queue of new items discovered. You can also manually add a new item if you know the IP address.
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.daikinairbase/README.md]:
```suggestion
This binding broadcasts messages on the local network looking for Daikin Airbase air conditioning units and adding them to the Inbox.
```

File: bundles/org.openhab.binding.daikinairbase/src/main/java/org/openhab/binding/daikinairbase/internal/handler/DaikinAirbaseAcUnitHandler.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+    private DaikinAirbaseWebTargets webTargets;
+    private ScheduledFuture<?> pollFuture;
+
+    public DaikinAirbaseAcUnitHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.daikinairbase/src/main/java/org/openhab/binding/daikinairbase/internal/handler/DaikinAirbaseAcUnitHandler.java]: You should also handle REFRESH commands within this method.

File: bundles/org.openhab.binding.daikinairbase/README.md (hunk 1 above)
Comment (psmedley) [reply bundles/org.openhab.binding.daikinairbase/README.md]: I believe 7a1a76b50d fixes these comments on README.md

File: bundles/org.openhab.binding.daikinairbase/src/main/java/org/openhab/binding/daikinairbase/internal/handler/DaikinAirbaseAcUnitHandler.java (hunk 2 above)
Comment (psmedley) [reply bundles/org.openhab.binding.daikinairbase/src/main/java/org/openhab/binding/daikinairbase/internal/handler/DaikinAirbaseAcUnitHandler.java]: aebd553945 should add support for REFRESH

**Comment by psmedley** [comment]
> 
> 
> This might be too broad. Please note that a REFRESH command is sent per channel - upon startup, you might receive one for every linked channel and you most likely don't want to call `poll()` 10 times in a row.
> As a solution, we often use the `ExpiringCache` class through which you can return a current value and which only calls poll, if the cached value is expired (expiration length should be the minimal frequency that you want `poll` to be called).

OK I'll do some investigation (hopefully tonight). Do you know a sample binding that uses the above approach? If not, I'll grep the code.

**Comment by kaikreuzer** [comment]
I don't have one at hand, would have to search as well, sorry...

**Comment by psmedley** [comment]
No worries, I'll try look tonight!

Cheers,

Paul

**Comment by kaikreuzer** [comment]
Sorry @psmedley, I had missed your last comment. As it is hardly readable here on github, could you maybe point to a branch in your fork with your current code, which I could checkout and see where the problem is?

**Comment by psmedley** [comment]
@kaikreuzer I'll set something up tonight

**Comment by psmedley** [comment]
@kaikreuzer enough to show the build error is at https://github.com/psmedley/openhab2-addons/tree/daikinairbasetests

I'm sure it's my lack of Java skills.....

**Comment by kaikreuzer** [comment]
@psmedley You seemed to not yet have grasped the idea of the `ExpiringCache`. The second parameter has to be a `Supplier`, i.e. a method, which is able to return the current value(s).

Please check the documentation at https://www.eclipse.org/smarthome/documentation/features/frameworkUtilities.html#caching - this should help you!

P.S.: Just noticed when checking out your branch that the binding is lacking the `.classpath` and `.project` files for the IDE. Please add them.

**Comment by psmedley** [comment]
> 
> 
> @psmedley You seemed to not yet have grasped the idea of the `ExpiringCache`. The second parameter has to be a `Supplier`, i.e. a method, which is able to return the current value(s).
> 
> Please check the documentation at https://www.eclipse.org/smarthome/documentation/features/frameworkUtilities.html#caching - this should help you!
> 
> P.S.: Just noticed when checking out your branch that the binding is lacking the `.classpath` and `.project` files for the IDE. Please add them.

OK I think I understand - so i need to cache something like getControlInfo() from DaikinAirbaseWebTargets.java that returns a structure with variables

**Comment by psmedley** [comment]
ok, https://github.com/psmedley/openhab2-addons/tree/daikinairbasetests now compiles - still need to study other examples to make sure I'm doing this right...

**Comment by kaikreuzer** [comment]
Sounds good. Let me know once you ar confident it works and have updated this PR.

**Comment by Hilbrand** [comment]
I'm not sure if it's a good idea to create this as a separate binding. It's basically a copy of the Daikin binding with modifications for a different api. If I understand it correctly that means 2 bindings supporting the same devices only with a different api. So any feature in this binding added might also be interesting in the original Daikin binding. So it would be better to have the device as a separate thing or just a configuration option if the devices are the same, so code can be shared and only differ on API interfacing. What was the reason to copy instead of enhancing the daikin binding?

**Comment by psmedley** [comment]
> I'm not sure if it's a good idea to create this as a separate binding. It's basically a copy of the Daikin binding with modifications for a different api. If I understand it correctly that means 2 bindings supporting the same devices only with a different api. So any feature in this binding added might also be interesting in the original Daikin binding. So it would be better to have the device as a separate thing or just a configuration option if the devices are the same, so code can be shared and only differ on API interfacing. What was the reason to copy instead of enhancing the daikin binding?

The two bindings support different devices sold into different markets. I initially tried to enhance the daikin binding but struggled to successfully modify it to support the two different APIs and to detect which type of device it was and adjust the code accordingly.

**Comment by psmedley** [comment]
> I'm not sure if it's a good idea to create this as a separate binding. It's basically a copy of the Daikin binding with modifications for a different api. If I understand it correctly that means 2 bindings supporting the same devices only with a different api. So any feature in this binding added might also be interesting in the original Daikin binding. So it would be better to have the device as a separate thing or just a configuration option if the devices are the same, so code can be shared and only differ on API interfacing. What was the reason to copy instead of enhancing the daikin binding?

Hi - https://github.com/psmedley/openhab2-addons/commit/bc13250a2b5e487d1ce19b7916d7e1b19e643e8d is a 1st attempt to merge support for the Airbase into the daikin binding. I still need to update readme.md....

**Comment by psmedley** [comment]
Closing this PR ie lieu of enhancing the existing daikin binding to support the Airbase.

**Pull request closed without merging**